File: 496-next-greater-element-i/next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int, int> nextgreater;
        stack<int> st;
        for (int num : nums2) {
            while (!st.empty() && st.top() < num) {
                nextgreater[st.top()] = num;
                st.pop();
            }
            st.push(num);
        }
        vector<int> res;
        for (int num : nums1) {
            res.push_back(nextgreater.count(num) ? nextgreater[num] : -1);
        }
        return res;
    }
};

File: 496-next-greater-element-i/next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        nextgreater = {}
        st = []
        for num in nums2:
            while st and st[-1] < num:
                nextgreater[st.pop()] = num
            st.append(num)
        return [nextgreater.get(num, -1) for num in nums1]
